# Lab 13: Gaussian Elimination with Backward Substitution/Gaussian Elimination with Scaled Partial Pivoting

### Balancing a chemical equation often requires solving a linear system of algebraic equations. Consider the chemical equation

$$
x_1 \text{K}_4\text{Fe}(\text{CN})_6 + x_2 \text{KMnO}_4 + x_3 \text{H}_2\text{SO}_4 \rightarrow x_4 \text{KHSO}_4 + 5\text{Fe}_2(\text{SO}_4)_3 + x_5 \text{MnSO}_4 + x_6 \text{HNO}_3 + x_7 \text{CO}_2 + x_8 \text{H}_2\text{O}
$$

(a) Since mass is conserved in a chemical reaction, the numbers and types of atoms should be the same before and after the reaction. Write down equations for $ x_1, x_2, \ldots, x_8 $ that reflect the previous statement. (Hint: For each type of atom, write an equation that represents the conservation of the numbers of this atom.)

(b) Move all the unknowns to the left-hand side of the equations and move all constants to the right-hand side. Write down the augmented matrix of this system.

(c) Use your Python function `gaussianElim` to find the solution of the system in (b), and write down the balanced chemical equation.


a)

for K: $4x_1 + x2 = x_4$

for Fe: $x_1 = 10$

for C: $6x_1 = x_7$

for N: $6x_1 = x_6$

for Mn: $x_2 = x_5$

for O: $4x_2 + 4x_3 = 4x_4 + 5(3)(4) + 4x_5 + 3x_6 + 2x_7 + x_8$

for H: $2x_3 = x_4 + x_6 + 2x_8$

for S: $x_3 = x_4 + 5(3) + x_5$

b)

The augmented matrix $ A $ is:

$$
A = \begin{bmatrix}
4 & 1 & 0 & -1 & 0 & 0 & 0 & 0 & | & 0 \\
1 & 0 & 0 & 0 & 0 & 0 & 0 & 0 & | & 10 \\
6 & 0 & 0 & 0 & 0 & 0 & -1 & 0 & | & 0 \\
6 & 0 & 0 & 0 & 0 & -1 & 0 & 0 & | & 0 \\
0 & 1 & 0 & 0 & -1 & 0 & 0 & 0 & | & 0 \\
0 & 4 & 4 & -4 & -4 & -3 & -2 & -1 & | & 60 \\
0 & 0 & 2 & -1 & 0 & -1 & 0 & -2 & | & 0 \\
0 & 0 & 1 & -1 & -1 & 0 & 0 & 0 & | & 15 \\
\end{bmatrix}
$$



In [1]:
# c

from NumericalMethodsCode.gaussianElim import gaussianElim
import numpy as np

A = np.array([
    [4., 1., 0., -1., 0., 0., 0., 0., 0.],
    [1., 0., 0., 0., 0., 0., 0., 0., 10.],
    [6., 0., 0., 0., 0., 0., -1., 0., 0.],
    [6., 0., 0., 0., 0., -1., 0., 0., 0.],
    [0., 1., 0., 0., -1., 0., 0., 0., 0.],
    [0., 4., 4., -4., -4., -3., -2., -1., 60.],
    [0., 0., 2., -1., 0., -1., 0., -2., 0.],
    [0., 0., 1., -1., -1., 0., 0., 0., 15.]
])

x = gaussianElim(A)
np.set_printoptions(precision=16)
print(f"x = ")
print(f"{x}")


x = 
[ 10. 122. 299. 162. 122.  60.  60. 188.]


$$
10 \, \text{K}_4\text{Fe}(\text{CN})_6 + 122 \, \text{KMnO}_4 + 299 \, \text{H}_2\text{SO}_4 \rightarrow 162 \, \text{KHSO}_4 + 5 \, \text{Fe}_2(\text{SO}_4)_3 + 122 \, \text{MnSO}_4 + 60 \, \text{HNO}_3 + 60 \, \text{CO}_2 + 188 \, \text{H}_2\text{O}
$$


### Consider the linear system

$$
3.3330x_1 + 15920x_2 + 10.333x_3 = 7953 \\
2.2220x_1 + 16.710x_2 + 9.6120x_3 = 0.965 \\
-1.5611x_1 + 5.1792x_2 - 1.6855x_3 = 2.714
$$

(a) Use both functions `gaussianElim` and `gaussianElimScaledPivot` to solve the system. Use the `numpy.set_printoptions(precision=16)` command in NumPy to print your results.

(b) Why are there differences in the solutions obtained by the two functions? Which one is more accurate?


In [4]:
# a
from NumericalMethodsCode.gaussianElimScaledPivot import gaussianElimScaledPivot

A1 = np.array([
    [3.3330, 15920, 10.333, 7953],
    [2.2220, 16.710, 9.6120, 0.965],
    [-1.5611, 5.1792, -1.68855, 2.714]
])
B1 = np.array(A1)
x1 = gaussianElim(A1)
np.set_printoptions(precision=16)
print(f"x1 = ")
print(f"{x1}")
y1 = gaussianElimScaledPivot(B1)
print(f"y1 =")
print(f"{y1}")

x1 = 
[ 1.0026040758642436  0.4999998453595026 -1.000601713746069 ]
y1 =
[ 1.0026040758644699  0.4999998453595028 -1.0006017137461656]


In [11]:
import numpy as np

# Define the system coefficients and RHS values
coefficients = np.array([
    [3.3330, 15920, 10.333],
    [2.2220, 16.710, 9.6120],
    [-1.5611, 5.1792, -1.6855]
])
rhs_values = np.array([7953, 0.965, 2.714])

# Solutions x1 and y1
x1 = np.array([1.0026040758642436, 0.4999998453595026, -1.000601713746069])
y1 = np.array([1.0026040758644699, 0.4999998453595028, -1.0006017137461656])

# Compute RHS values for x1 and y1 solutions
rhs_x1 = np.dot(coefficients, x1)
rhs_y1 = np.dot(coefficients, y1)

# Compute absolute differences between computed RHS values and original RHS
diff_x1 = np.abs(rhs_values - rhs_x1)
diff_y1 = np.abs(rhs_values - rhs_y1)

print(f"Absolute error for gaussianElim: {diff_x1}") 
print(f"Absolute error for gaussianElimScaledPivot: {diff_y1}")


Absolute error for gaussianElim: [9.0949470177292824e-13 4.2288395007972213e-13 3.0518352267354665e-03]
Absolute error for gaussianElimScaledPivot: [1.8189894035458565e-12 7.7715611723760958e-16 3.0518352269250926e-03]


Scaled pivot version picked the larger scaled quantity in the column as the pivot, so scaled pivoting is more accurate.